src/wavm.cpp

```diff
@@ -0,0 +1,183 @@
+    return result;
+  }
+
+
+  DEFINE_INTRINSIC_FUNCTION_WITH_MEM_AND_TABLE(ethereum, "useGas", void, useGas, I64 amount)
+  {
+    printf("useGas  amount=%lu\n",amount);
+    interface.top()->eeiUseGas(amount);
+  }
+
+
+  // this is needed for resolving names of imported host functions
+  struct MyResolver : Runtime::Resolver {
+    Runtime::Compartment* compartment;
+    HashMap<std::string, Runtime::ModuleInstance*> moduleNameToInstanceMap;
+
+    MyResolver(Runtime::Compartment* inCompartment) : compartment(inCompartment) {}
+
+    bool resolve(const std::string& moduleName,
+           const std::string& exportName,
+           IR::ObjectType type,
+           Runtime::Object*& outObject) override {
+      auto namedInstance = moduleNameToInstanceMap.get(moduleName);
+      outObject = getInstanceExport(*namedInstance, exportName);
+      return true;
+    }
+  };
+
+} //namespace wavm_host_module
+
+
+
+ExecutionResult WavmEngine::execute(
+  evmc_context* context,
+  vector<uint8_t> const& code,
+  vector<uint8_t> const& state_code,
+  evmc_message const& msg,
+  bool meterInterfaceGas
+) {
+  HERA_DEBUG << "Executing with wavm...\n";
+
+  // set up a new ethereum interface just for this contract invokaction
+  ExecutionResult result;
+  wavm_host_module::interface.push(new WavmEthereumInterface{context, state_code, msg, result, meterInterfaceGas});
+
+  // first parse module
+  std::vector<U8> codeBytes = static_cast< std::vector<U8> > (code);
+  IR::Module moduleAST; 
+  loadBinaryModule(codeBytes.data(), codeBytes.size(), moduleAST);
+
+  // next set up the host module. Note: in ewasm, we create a new VM for each call to a module, so we must instantiate a new host module for each of these VMs, this is inefficient, but OK for prototyping.
+  // compartment is like the Wasm store, represents the VM, has lists of globals, memories, tables, and also has wavm's runtime stuff
+  Runtime::Compartment* compartment = Runtime::createCompartment();
+  // context stores the compartment and some other stuff
+  Runtime::Context* wavm_context = Runtime::createContext(compartment);
+  // instantiate host Module
+  HashMap<std::string, Runtime::Object*> extraEthereumExports; //empty for current ewasm stuff
+  Runtime::GCPointer<Runtime::ModuleInstance> ethereumHostModule;
+  ethereumHostModule = Intrinsics::instantiateModule(compartment, wavm_host_module::INTRINSIC_MODULE_REF(ethereum), "ethereum", extraEthereumExports);
+  // prepare contract module to resolve links against host module
+  wavm_host_module::MyResolver myResolver(compartment);
+  myResolver.moduleNameToInstanceMap.set("ethereum", ethereumHostModule);
+  Runtime::LinkResult linkResult = linkModule(moduleAST, myResolver);
+
+  // instantiate contract module
+  Runtime::ModuleInstance* moduleInstance = instantiateModule( compartment, moduleAST, std::move(linkResult.resolvedImports), "arbitrary_name_of_contract_module");
```

Comment:
Can you really move `resolvedImports` here?

Reply:
What is `"arbitrary_name_of_contract_module"` exactly for?

Reply:
I will retest `resolvedImports`, don't remember why it is done like this, but it worked in early tests.

`"arbitrary_name_of_contract_module"` is the name of the module, arbitrary for now. Some people prefer things like this to document themselves, but I will change it to the empty string.